src/storage/archive_index.rs

```diff
@@ -0,0 +1,106 @@
+use crate::storage::{compression::CompressionAlgorithm, FileRange};
+use failure::Error;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::io;
+use std::path::{Path, PathBuf};
+
+#[derive(Deserialize, Serialize)]
+pub struct FileInfo {
+    range: FileRange,
+    compression: CompressionAlgorithm,
+}
+
+impl FileInfo {
+    pub fn range(&self) -> FileRange {
+        self.range.clone()
+    }
+    pub fn compression(&self) -> CompressionAlgorithm {
+        self.compression
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct Index {
+    files: HashMap<PathBuf, FileInfo>,
+}
+
+impl Index {
+    pub fn load(reader: impl io::Read) -> Result<Index, Error> {
+        serde_json::from_reader(reader).map_err(|_| failure::err_msg("deserialization error"))
+    }
+
+    pub fn save(&self, writer: impl io::Write) -> Result<(), Error> {
+        serde_json::to_writer(writer, self).map_err(|_| failure::err_msg("serialization error"))
+    }
+
+    pub fn new_from_zip<R: io::Read + io::Seek>(zipfile: &mut R) -> Result<Index, Error> {
+        let mut archive = zip::ZipArchive::new(zipfile)?;
+
+        // get file locations
+        let mut files: HashMap<PathBuf, FileInfo> = HashMap::new();
+        for i in 0..archive.len() {
+            let zf = archive.by_index(i)?;
+
+            files.insert(
+                PathBuf::from(zf.name()),
+                FileInfo {
+                    range: zf.data_start()..=(zf.data_start() + zf.compressed_size() - 1),
```

Comment:
that doesn't work (directly) because `range` is `ops::RangeInclusive`.

I didn't directly see a simple way to accept more ranges, but of course I might miss something. I didn't didn't want to make the struct generic only for the sake of accepting `ops::Range` here. 

If there is a better way, I'm happy to implement it :)

Reply:
Ok, I see https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.35 says HTTP ranges are inclusive. The current code seems fine then.